### ISSUE TITLE
Add labels on dependency dashboard issue depending on its content

### ISSUE THREAD

**Issue by scrocquesel** [renovatebot/renovate#12414]
### What would you like Renovate to be able to do?

When having a lot of project, it can be painful to open each dependency dashboard issue to know if manual action are required.

I was first thinking of having a label for when there is a repo problem or pending approval, but maybe other sections can have there own label.

This issue can try to list which labels can be helpfull and how to name the configuration property ? (Maybe is it time to have a dedicated configuration object because this can add a lot of dependencyDashboardLabelsxxx.)

Here the list I extract from 

- [x] repoProblems (Repository problems)
- [x] pendingApprovals (Pending Approval)
- [ ] awaitingSchedule (Awaiting Schedule)
- [ ] rateLimited (Rate Limited)
- [ ] errorList (Errored) (could it share its label with repoProblems ?)
- [x] awaitingPr (PR Creation Approval Required)
- [x] prEdited (Edited/Blocked)
- [ ] prPending (Pending Status Checks)
- [ ] prPendingBranchAutomerge (Pending Branch Automerge)
- [x] otherBranches (Other Branches)
- [ ] inProgress (Open)
- [x] alreadyExisted (Ignored or Blocked)

### If you have any ideas on how this should be implemented, please tell us here.

All sections are built in https://github.com/renovatebot/renovate/blob/1f86b71fd5f9175cdb6ab8c673dbdf83082c2c2e/lib/workers/repository/dependency-dashboard.ts#L98, so we can collect labels while building the body and merge them with config.dependencyDashboardLabels.

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by astellingwerf** [comment]
I like the idea! I do see a slight potential issue/confusion with `prPendingBranchAutomerge`, when relying on the platform handling the automerging.

Would it make sense to have a package rule to decide on the label on the dashboard? I can imagine scenarios where different package results would be reason for different labels to be added to the dashboard.

**Comment by scrocquesel** [comment]
@astellingwerf I think that package result is unaware of the state of the PR. We can already add label on the PR with package rule based on Package Results, so they are searchable.
Maybe we should be able to add label on the PR based on their state. IMHO, the only state it can be usefull is `prEdited`. This can be tracked in an another issue.

I'm not sure for `otherBranches` as I don't really understand when it happens.

Then, we are left with states/errors that are not associated with a PR : `repoProblems`, `errorList`,  `pendingApprovals`.

To begin with, what about adding two options `dependencyDashboardLabelsOnError` and `dependencyDashboardLabelsOnPendingApproval` ?

```js
  {
    name: 'dependencyDashboardLabelsOnError',
    description:
      'These labels will be applied on the Dependency Dashboard issue when an error occured while renovating the repository, even when they have been removed manually.',
    type: 'array',
    subType: 'string',
    default: null,
  },
  {
    name: 'dependencyDashboardLabelsOnPendingApproval',
    description:
      'These labels will be applied on the Dependency Dashboard issue when updates require approval, even when they have been removed manually.',
    type: 'array',
    subType: 'string',
    default: null,
  }
```

@rarkins WDYT ?

**Comment by rarkins** [comment]
I'm ok with the idea if it can be implemented elegantly. Need to consider what empty array [] means. I'd prefer it means the same as null, and could be used as default instead.

Also, should these be top-level options only? E.g. not within packageRules

**Comment by scrocquesel** [comment]
> I'm ok with the idea if it can be implemented elegantly. Need to consider what empty array [] means. I'd prefer it means the same as null, and could be used as default instead.
> 

I use the same design as the existing `dependencyDashboardLabels`. Whatever the default value is, the three options should be aligned.

> Also, should these be top-level options only? E.g. not within packageRules

Repo errors may not be hit with package rules as it can result in no package or update. It may be interesting for package lookup errors (e.g. authent issues) but I don't see why only some errors should be notified by a label. E.g. npm dev dependencies but not npm dependencies. If you don't mind one error, it seems you don't mind update and you may disable renovate with a package rule for these one.

For pending approval, the idea behind the dashboard issue is to make them visible, so does the `dependencyDashboardLabelsOnPendingApproval`.

To keep it simple, these two labels are very closed to the content of the dashboard issue. If a section is present, then a corresponding label should help find out when, because they require an action from the user and no PR can help to guess that.

**Comment by scrocquesel** [comment]
If we can have a rule object with at least `matchBranchResult` and `dependencyDashboardLabels`, it could solve the issue (apart from repo problems that still need a top-level options).

Assuming we can use PackageRule to do so, use cases could be :

```jsonc
[
  { // if at least one branch 
    "matchManagers": ["dockerfile"], // is an update managed by dockerfile, 
    "dependencyDashboardLabels": [ "docker" ] // then dashboard issue will have the label "docker"
  },
  { // if at least one branch 
    "matchManagers": ["dockerfile"], // is an update managed by dockerfile
    "matchBranchResult": "needs-pr-approval", // AND need a pr approval
    "dependencyDashboardLabels": [ "contains-pr-approval" ],  // then dashboard issue will have the label "contains-pr-approval"
    "labels": [ "not used" ] // the PR doesn't exist, labels has no effect
  },
  { // if at least one branch
    "matchBranchResult": "error", // is in error
    "dependencyDashboardLabels": [ "contains-error" ]  // then dashboard issue will have the label "contains-error",
    "labels": [ "error" ] // then if a PR exists, it will have the label "error"
  },
  { // if at least one branch
    "matchBranchResult": "pr-edited",  // has been edited
    "dependencyDashboardLabels": [ "contains-pr-edited" ], // then dashboard issue will have the label "contains-pr-edited"
    "labels": [ "pr-edited" ] // then the PR will have the label "pr-edited"
  }
]
```

**Comment by rarkins** [comment]
Another related point: users have different expectations on what dependency dashboard labels means:
- Is it labels that are added on issue creation, and then never again modified after it exists?
- Is it labels that Renovate should ensure are always there (meaning that if you modify your config, or remove labels, you'll see Renovate add the labels to the issue). It means that if you had label "foo" and changed to label "bar", Renovate will add foo but won't delete bar
- Is it exclusive labels? Renovate will enforce these labels and delete anything else which a user adds

We maybe need to make this configurable too!

**Comment by scrocquesel** [comment]
> Another related point: users have different expectations on what dependency dashboard labels means:
> 
> * Is it labels that are added on issue creation, and then never again modified after it exists?
> * Is it labels that Renovate should ensure are always there (meaning that if you modify your config, or remove labels, you'll see Renovate add the labels to the issue). It means that if you had label "foo" and changed to label "bar", Renovate will add foo but won't delete bar
> * Is it exclusive labels? Renovate will enforce these labels and delete anything else which a user adds
> 
> We maybe need to make this configurable too!

Currently, for dependency dashboard labels, Renovate ensure the configured labels and only them are set on the issue (it replaces all labels with the one configured when the dashboard board issue is updated (which may not be on each run)).
The purpose of this issues is to reflect the current content of the dashboard issue as label. Therefore, the only meaningfull option is to replace labels too.
Therefore, label computed from the rules shall be added to the dependencyDashboardLabels and the resulting set replacing labels.

**Comment by rarkins** [comment]
Good point, this functionality wouldn't work unless we also remove labels

**Comment by scrocquesel** [comment]
I'm stuck trying to implement this as it become more complex than I first thought. Package rules cannot be applied to `BranchConfig`. Thus, we need to reimplement the flattenning logic with `BranchConfig`.
We can tried to flow the matching package rules on the Branch Config during package analysis but the current logic will not be able to match the two last use case . Unless it knows there is some options on the package rule it cannot evaluate right now and if all other options are valid or no other one exist, it will consider the package rule as matching. This could lead to configuration behaving not as expected.

Maybe we should split Package Rule and Branch Rule ?

**Comment by rarkins** [comment]
No to the concept of branch rules

**Comment by scrocquesel** [comment]
While the feature still interest me, it implies too much modification and know-how around configuration I can handle by myself
